vespajlib/src/main/java/com/yahoo/tensor/TensorDataSource.java

```diff
@@ -0,0 +1,288 @@
+            emitValue(tensor.get(0), tensor.type().valueType(), sink);
+        } else {
+            for (indexes[dimension] = 0; indexes[dimension] < sizes.size(dimension); ++indexes[dimension]) {
+                if (dimension < (sizes.dimensions() - 1)) {
+                    sink.startArray();
+                    emitDenseValuesRecursive(tensor, sink, indexes, dimension + 1);
+                    sink.endArray();
+                } else {
+                    emitValue(tensor.get(indexes), tensor.type().valueType(), sink);
+                }
+            }
+        }
+    }
+
+    private void emitLabeledBlocks(MixedTensor tensor, DataSink sink)  {
+        sink.startObject();
+        TensorType denseSubType = tensor.type().indexedSubtype();
+        for (var subspace : tensor.getInternalDenseSubspaces()) {
+            String label = subspace.sparseAddress.label(0);
+            sink.fieldName(label);
+            if (options.hexForDensePart()) {
+                sink.stringValue(asHexString(subspace.cells.length, denseSubType.valueType(),
+                                             i -> subspace.cells[i], i -> (float) subspace.cells[i]));
+            } else {
+                IndexedTensor denseSubspace = IndexedTensor.Builder.of(denseSubType, subspace.cells).build();
+                emitDenseValues(denseSubspace, sink);
+            }
+        }
+        sink.endObject();
+    }
+
+    private void emitAddressedBlocks(MixedTensor tensor, DataSink sink)  {
+        sink.startArray();
+        var mappedDimensions = tensor.type().dimensions().stream()
+                .filter(TensorType.Dimension::isMapped)
+                .toList();
+        TensorType denseSubType = tensor.type().indexedSubtype();
+
+        for (var subspace : tensor.getInternalDenseSubspaces()) {
+            sink.startObject();
+
+            sink.fieldName("address");
+            sink.startObject();
+            for (int i = 0; i < mappedDimensions.size(); i++) {
+                sink.fieldName(mappedDimensions.get(i).name());
+                sink.stringValue(subspace.sparseAddress.label(i));
+            }
+            sink.endObject();
+
+            sink.fieldName("values");
+            if (options.hexForDensePart()) {
+                sink.stringValue(asHexString(subspace.cells.length, denseSubType.valueType(),
+                                             i -> subspace.cells[i], i -> (float) subspace.cells[i]));
+            } else {
+                IndexedTensor denseSubspace = IndexedTensor.Builder.of(denseSubType, subspace.cells).build();
+                emitDenseValues(denseSubspace, sink);
+            }
+
+            sink.endObject();
+        }
+        sink.endArray();
+    }
+
+    private void emitValue(double value, TensorType.Value valueType, DataSink sink)  {
+        switch (valueType) {
+            case DOUBLE:
+            case FLOAT:
+            case BFLOAT16:
+                sink.doubleValue(value);
```

Comment:
Could we call sink.floatValue for float/bfloat16 and sink.byteValue for int8?

Reply:
will do